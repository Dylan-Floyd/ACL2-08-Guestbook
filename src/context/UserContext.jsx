import { createContext, useContext, useState } from 'react'

const UserContext = createContext({})

export const UserProvider = ({ children }) => {
  const [user, setUser] = useState({})

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  )
}

export const useUser = () => {
  const context = useContext(UserContext)
  
  if (context === undefined) {
    throw new Error('useUser can only be called from a child component of UserProvider')
  }

  return context;
}