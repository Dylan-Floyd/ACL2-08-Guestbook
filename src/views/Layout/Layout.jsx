import { useUser } from '../../context/UserContext.jsx'

export default function Layout({ children }) {
  const { user } = useUser()

  return (
    <>
      <header>
        <h2>Guestbook</h2>
        <div className='right-div'>
          {user?.name ? <>(Signing as {user.name})</> : <></>}
        </div>
      </header>
      {children}
      <footer>
        Footer
      </footer>
    </>
  )
}
