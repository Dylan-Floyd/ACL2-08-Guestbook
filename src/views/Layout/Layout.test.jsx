import { 
  screen,
  render,
  waitForElementToBeRemoved,
  within,
  act,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserProvider, useUser } from '../../context/UserContext.jsx';
import Layout from './Layout.jsx';

let outerSetUser

function SignInMocker() {
  const { user, setUser } = useUser()

  outerSetUser = setUser;

  return <></>
}

it('Renders the header, footer, and children, and also the username if signed in', async () => {
  const { container } = render(
    <UserProvider>
      <Layout>
        <h1>I'm a child</h1>
        <SignInMocker />
      </Layout>
    </UserProvider>
  )

  const header = screen.getByText('Guestbook')
  const footer = screen.getByText('Footer')

  //set a timeout so the next while loop won't wait forever
  const timer = setTimeout(() => {
    throw new Error('The SignInMocker didn\'t work')
  }, 1000);

  //wait until this gets set
  while(outerSetUser === undefined) {}
  clearTimeout(timer)

  act(() => outerSetUser({ name: 'bob'}))

  const bob = await screen.findByText(/bob/)
  expect(bob).toBeInTheDocument()

  expect(container).toMatchSnapshot()
})