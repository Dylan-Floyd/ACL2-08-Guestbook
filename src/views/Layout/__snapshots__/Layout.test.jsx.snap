// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the header, footer, and children, and also the username if signed in 1`] = `
<div>
  <header>
    <h2>
      Guestbook
    </h2>
    <div
      class="right-div"
    >
      (Signing as 
      bob
      )
    </div>
  </header>
  <h1>
    I'm a child
  </h1>
  <footer>
    Footer
  </footer>
</div>
`;
